### PR TITLE
#2051 Make kv index for tx.height

### PR DESCRIPTION
**Pull request by ackratos**
Partially fixed https://github.com/tendermint/tendermint/issues/2051 to make rest api request like http://127.0.0.1:8080/txs?tag=tx.height=1330 work

This is tested against our ABCI application which relies on cosmos default Send msg tx.
To make sure this change works, I:
0. change config.toml file: `index_all_tags=true` and `timeout_commit=30000` (so that 2 transactions issued manually can be included in same block
1. started local test net in 4 docker container
2. restored two accounts for different home (so that I can run 2 gaiacli to make transaction at the same time)
3. transfer coins "at the same time"
4. start a node rest api-server and check **http://127.0.0.1:8080/txs?tag=tx.height=1330**

I can get json response like:

```
[  
   {  
      hash:"798E4CE82A5CFEB65E9DECD3F4D00F9753509FE7",
      height:"1330",
      tx:{  
         type:"auth/StdTx",
         value:{  
            msg:[  
               {  
                  type:"cosmos-sdk/Send",
                  value:{  
                     inputs:[  
                        {  
                           address:"cosmosaccaddr19zqnmalreaugjf8mxm5z0x99ew49anghxpv3x4",
                           coins:[  
                              {  
                                 denom:"BBB",
                                 amount:"10000000000"
                              }
                           ]
                        }
                     ],
                     outputs:[  
                        {  
                           address:"cosmosaccaddr1gtj97w4sxz0eademencfdj8nesmy4c306a3g50",
                           coins:[  
                              {  
                                 denom:"BBB",
                                 amount:"10000000000"
                              }
                           ]
                        }
                     ]
                  }
               }
            ],
            fee:{  
               amount:[  
                  {  
                     denom:"",
                     amount:"0"
                  }
               ],
               gas:"200000"
            },
            signatures:[  
               {  
                  pub_key:{  
                     type:"tendermint/PubKeySecp256k1",
                     value:"AgOIKfLYQ0qmtER5Dy58oH4VcAEJodb4xMG29QuHr/2P"
                  },
                  signature:{  
                     type:"tendermint/SignatureSecp256k1",
                     value:"MEUCIQCJSAW5yPYpiHLeSm4DwnQUrwXz05sYB1cD10kS4o3qSAIgQu/a7+WS3fh6KCQ8MMML+Q4sYVDH4aa3nPzvrPLbGyg="
                  },
                  account_number:"2",
                  sequence:"0"
               }
            ],
            memo:""
         }
      },
      result:{  
         log:"Msg 0: ",
         gasUsed:"4924",
         tags:[  
            {  
               key:"c2VuZGVy",
               value:"Y29zbW9zYWNjYWRkcjE5enFubWFscmVhdWdqZjhteG01ejB4OTlldzQ5YW5naHhwdjN4NA=="
            },
            {  
               key:"cmVjaXBpZW50",
               value:"Y29zbW9zYWNjYWRkcjFndGo5N3c0c3h6MGVhZGVtZW5jZmRqOG5lc215NGMzMDZhM2c1MA=="
            }
         ],
         fee:{  

         }
      }
   },
   {  
      hash:"81D9D88B85B86C52BEEAE56CF77791AEE81A08FE",
      height:"1330",
      tx:{  
         type:"auth/StdTx",
         value:{  
            msg:[  
               {  
                  type:"cosmos-sdk/Send",
                  value:{  
                     inputs:[  
                        {  
                           address:"cosmosaccaddr1gtj97w4sxz0eademencfdj8nesmy4c306a3g50",
                           coins:[  
                              {  
                                 denom:"BBB",
                                 amount:"100000000000"
                              }
                           ]
                        }
                     ],
                     outputs:[  
                        {  
                           address:"cosmosaccaddr19zqnmalreaugjf8mxm5z0x99ew49anghxpv3x4",
                           coins:[  
                              {  
                                 denom:"BBB",
                                 amount:"100000000000"
                              }
                           ]
                        }
                     ]
                  }
               }
            ],
            fee:{  
               amount:[  
                  {  
                     denom:"",
                     amount:"0"
                  }
               ],
               gas:"200000"
            },
            signatures:[  
               {  
                  pub_key:{  
                     type:"tendermint/PubKeySecp256k1",
                     value:"Aso+s1hyS8mhjVxH0TsVtEKe7jyq5XHY6e4UhnYXqWbI"
                  },
                  signature:{  
                     type:"tendermint/SignatureSecp256k1",
                     value:"MEQCIFRRELdb6gMV6ICRUXIF4BVCiKXiSEnZ2oGaQKkGeKZpAiA76pVmlFSlGco6O6cbMIcylq3KijF7dxY0e74HcikZvg=="
                  },
                  account_number:"0",
                  sequence:"2"
               }
            ],
            memo:""
         }
      },
      result:{  
         log:"Msg 0: ",
         gasUsed:"5029",
         tags:[  
            {  
               key:"c2VuZGVy",
               value:"Y29zbW9zYWNjYWRkcjFndGo5N3c0c3h6MGVhZGVtZW5jZmRqOG5lc215NGMzMDZhM2c1MA=="
            },
            {  
               key:"cmVjaXBpZW50",
               value:"Y29zbW9zYWNjYWRkcjE5enFubWFscmVhdWdqZjhteG01ejB4OTlldzQ5YW5naHhwdjN4NA=="
            }
         ],
         fee:{  

         }
      }
   }
]
```

* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [x] Updated CHANGELOG_PENDING.md

### CODE REVIEW

File: state/txindex/kv/kv.go [hunk 1]
```diff
@@ -89,6 +89,11 @@ func (txi *TxIndex) AddBatch(b *txindex.Batch) error {
 			}
 		}
 
+		// index tx by height
+		if txi.indexAllTags || cmn.StringInSlice("tx.height", txi.tagsToIndex) {
```
[the remark this answers is not in the document]
Comment (melekes) [reply state/txindex/kv/kv.go]: Could you also make the same changes in TxIndex#Index method?

File: state/txindex/kv/kv.go [hunk 2]
```diff
@@ -154,10 +159,7 @@ func (txi *TxIndex) Search(q *query.Query) ([]*types.TxResult, error) {
 	skipIndexes := make([]int, 0)
 
 	// if there is a height condition ("tx.height=3"), extract it for faster lookups
-	height, heightIndex := lookForHeight(conditions)
-	if heightIndex >= 0 {
-		skipIndexes = append(skipIndexes, heightIndex)
-	}
+	height, _ := lookForHeight(conditions)
```
Comment (melekes) [inline state/txindex/kv/kv.go]: Looks like we don't need heightIndex anymore. Can you modify `lookForHeight` to only return height then?

File: state/txindex/kv/kv.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ackratos) [reply state/txindex/kv/kv.go]: @melekes that make sense. I will refactor in next commit (amend)

File: state/txindex/kv/kv.go (hunk 2 above)
Comment (ackratos) [reply state/txindex/kv/kv.go]: @melekes sure, thanks. Will refactor in next commit

**Comment by ackratos** [comment]
> Could you add a changelog entry to CHANGELOG_PENDING.md file?

done~

**Comment by ebuchman** [comment]
THANKS!

**Comment by webmaster128** [comment]
Does this add support for range searches? E.g. `tx.height>100`, `tx.height<100`, `tx.height>=100`, `tx.height<=100`? Those seem to not work in Tendermint 0.25.0

**Pull request merged**